frontend/src/components/common/rich-text-editor/RichTextPasting.tsx

```diff
@@ -0,0 +1,487 @@
+
+  if (node.nextElementSibling) traverse(node.nextElementSibling, isNestedBlock)
+  isNestedBlock =
+    isNestedBlock || node.tagName === 'LI' || node.tagName === 'TD'
+  if (node.firstElementChild) traverse(node.firstElementChild, isNestedBlock)
+}
+
+/**
+ * Removes <style> tags if any
+ * @param parsedHtml html parsed by DOMParser
+ * @returns parsed html without style tags
+ */
+function removeStyleTags(parsedHtml: HTMLBodyElement): HTMLBodyElement {
+  let child = parsedHtml.firstChild as Element
+
+  while (child) {
+    if (child.tagName === 'STYLE') {
+      const nextChild = child.nextSibling as Element
+      parsedHtml.removeChild(child)
+      child = nextChild
+      continue
+    }
+    child = child?.nextSibling as Element
+  }
+  return parsedHtml
+}
+
+/**
+ * Cleans html
+ * - parse html using DOMParser
+ * - removes <style> tags if any
+ * - traverse the DOM tree, making modification to prepare for conversion to draft state
+ *    - refer to traverse function header
+ * - converts back to html string and adds a line break
+ * @param newHtml html of pasted text
+ * @returns cleaned html
+ */
+function cleanHtml(newHtml: string): string {
+  const domParser = new DOMParser()
+  const tempDoc = domParser.parseFromString(newHtml, 'text/html')
+  let parsedHTML = tempDoc.querySelector('body')
+  if (!parsedHTML) return ''
+
+  parsedHTML = removeStyleTags(parsedHTML)
+
+  if (parsedHTML.firstElementChild)
+    traverse(parsedHTML.firstElementChild, false)
+
+  const s = new XMLSerializer()
+  const cleanedHtml = s.serializeToString(parsedHTML).concat('<p></p>')
+
+  return cleanedHtml
+}
+
+/**
+ * Filter state to only keep items that our rich text editor supports
+ * Only render things that are controllable by our toolbar
+ * @param newEditorState editor state to filter
+ * @returns filtered state
+ */
+function filterState(newEditorState: EditorState): EditorState {
+  return filterEditorState(
+    {
+      blocks: [
+        'header-one',
+        'header-two',
+        'header-four',
+        'table-cell',
+        'ordered-list-item',
+        'unordered-list-item',
+      ],
+      styles: ['BOLD', 'ITALIC', 'UNDERLINE'],
+      entities: [
+        {
+          type: 'LINK',
+          attributes: ['url'],
+        },
+      ],
+      maxNesting: 5,
+      whitespacedCharacters: [],
+    },
+    newEditorState
+  )
+}
+
+/**
+ * Adjusts depth of list items
+ * - reads depth from data.level and sets block depth
+ * - data.level represents the item depth and was populated during custom block conversion
+ * @param newEditorState editor state to adjust depth
+ * @returns state with list depth adjusted
+ */
+function adjustListDepth(newEditorState: EditorState): EditorState {
+  const curContent = newEditorState.getCurrentContent()
+  const newBlocks = curContent.getBlocksAsArray()
+  let blockMap = curContent.getBlockMap()
+
+  // sets depth of each item block
```

Comment:
Instead of this comment (which is quite clear from the code below), a more effective one would be explaining why we're lowering the depth/level by 1

Reply:
added